text/0322-cdk-pipelines-updated-api.md

```diff
@@ -0,0 +1,325 @@
+  as may be appropriate.
+
+### How do approval workflows work?
+
+Every `Approval` step will be handed a reference to the workflow of the application
+being deployed, as well as to an empty "approval" workflow. It can then choose to
+add new actions and dependencies to any of those workflows.
+
+In the most common case, it will add things like "shell script" actions to the approval workflow; but it might also
+choose to add actions before "Deploy" (`Approval.securityChanges()`), or it might choose to add actions *in between*
+`CreateChangeSet` and `DeployChangeSet` pairs.
+
+The actions it adds may be generic (from level 2), translatable into any backend and so
+the approval workflow is reusable, or they may be backend-specific and only apply to
+one specific backend.
+
+```
+┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┐
+                                                   App
+│┌────────────────────────────────────────────────────┐          ┌───────────────────────────────────┐  │
+ │                       Deploy                       │          │              Approve              │
+││                                                    │          │                                   │  │
+ │ ┌────────────────────┐    ┌────────────────────┐   │          │                                   │
+││ │       Stack1       │    │       Stack2       │   │          │                                   │  │
+ │ │                    │    │                    │   │          │                                   │
+││ │  ┌────┐    ┌────┐  │    │  ┌────┐    ┌────┐  │   │─────────▶│                                   │  │
+ │ │  │ C  │───▶│ E  │  │───▶│  │ C  │───▶│ E  │  │   │          │                                   │
+││ │  └────┘    └────┘  │    │  └────┘    └────┘  │   │          │                                   │  │
+ │ │                    │    │                    │   │          │                                   │
+││ └────────────────────┘    └────────────────────┘   │          │                                   │  │
+ │                                                    │          │                                   │
+│└────────────────────────────────────────────────────┘          └───────────────────────────────────┘  │
+ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─
+```
+
+### How will asset publishing work?
+
+Similar to how approval workflows work, the asset publishing strategy can be configured
+with a callback which can manipulate the workflow as desired. This allows for easy switching
+between:
+
+* Prepublish all assets in individual CodeBuild projects
+* Prepublish all assets in one CodeBuild project
+* Publish assets just before each app deployment
+* Publish initial assets as usual but afterwards wait for ECR replication to finish
```

Comment:
1. Why do we need these degrees of freedom?
2. If these are real requirements from users, I rather we implement them as "real" features and not as open hooks in the API.

Generally speaking, let's try to avoid "callback APIs" - that's a very fragile contract, and very hard to undo if we get wrong because we can never tell what people are doing with it.

Reply:
I've also heard "trigger a StepFunctions workflow to do the asset publishing".

I definitely don't want to add support for everything on our end. There is lots of value in extensibility.

Reply:
Supporting random customizations is not a requirement and usually not a good API design practice as it oftentimes leads to leaky APIs that are very hard to evolve. If there is a concrete use case with merit, we should consider how to support it and expose the correct API for it.

As for triggering step functions, what's the use case? What's the rationale? Or is it just one user's preference.

I rather we design our API with the minimal surface area and expand it as new requirements emerge then offer an extensible API that we won't be able to back away from.